### PR TITLE
Simplify adding table functions

### PR DESCRIPTION
**Pull request by vlad-lyutenko**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java [hunk 1]
```diff
@@ -180,13 +177,20 @@ public ScalarFunctionImplementation getScalarFunctionImplementation(
...
-        }
-        if (functionHandle instanceof SequenceFunctionHandle) {
-            return getSequenceFunctionProcessorProvider();
-        }
+        // TODO somehow to inject tableFunctions here
+        // TODO TableFunctionHandle doesn't contains name anymore https://github.com/trinodb/trino/pull/17117
+        // need somehow to map ConnectorTableFunctionHandle to TableFunctionHandle/ConnectorTableFunction
+        ConnectorTableFunctionHandle connectorTableFunctionHandle = null;
 
+        if (connectorTableFunctionHandle instanceof AbstractConnectorTableFunctionExtended abstractConnectorTableFunctionExtended) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java:185]: remove `AbstractConnectorTableFunctionExtended` interface

let GFC get a list of TableFunctionProcessorProvider factories. the factories will be called in sequence and only one can return a TableFunctionProcessorProvider

perhaps
```
return  tableFunctionProcessorFactories.stream()
    .map(factory -> getTableFunctionProcessorProvider(factory))
   .flaMap(Optional::stream)
    .collect(toOptional())
   .orElse(null);
```

Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java]: @findepi big thanks, refactored

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * limitations under the License.
+ */
+package io.trino.metadata;
+
+import io.trino.spi.function.table.ConnectorTableFunctionHandle;
+import io.trino.spi.function.table.TableFunctionProcessorProvider;
+
+public interface TableFunctionProcessorProviderFactory
+{
+    TableFunctionProcessorProvider getTableFunctionProcessorProvider(ConnectorTableFunctionHandle functionHandle);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java:21]: Can we make it Optional?

anyway, add a javadoc explaining what it does and what it returns when `functionHandle` is not recognized

Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java:21]: Can we make it Optional?

anyway, add a javadoc explaining what it does and what it returns when `functionHandle` is not recognized

Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java:21]: Can we make it Optional?

anyway, add a javadoc explaining what it does and what it returns when `functionHandle` is not recognized

Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java:21]: Can we make it Optional?

anyway, add a javadoc explaining what it does and what it returns when `functionHandle` is not recognized

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.metadata;
+
+import io.trino.spi.function.table.ConnectorTableFunctionHandle;
+import io.trino.spi.function.table.TableFunctionProcessorProvider;
+
+public interface TableFunctionProcessorProviderFactory
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java:19]: Should this and ConnectorSplitSourceFactory be one interface?

File: core/trino-main/src/main/java/io/trino/operator/table/ExcludeColumns.java [hunk 4]
```diff
@@ -149,21 +150,29 @@ public TableFunctionAnalysis analyze(
...
+        @Override
+        public TableFunctionProcessorProvider getTableFunctionProcessorProvider(ConnectorTableFunctionHandle functionHandle)
         {
-            @Override
-            public TableFunctionDataProcessor getDataProcessor(ConnectorTableFunctionHandle handle)
-            {
-                return input -> {
-                    if (input == null) {
-                        return FINISHED;
+            if (functionHandle instanceof ExcludeColumnsFunctionHandle) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/table/ExcludeColumns.java:159]: invert `if` to reduce overall code indentation

File: core/trino-main/src/main/java/io/trino/operator/table/Sequence.java [hunk 5]
```diff
@@ -236,16 +251,24 @@ public long getRetainedSizeInBytes()
...
     {
-        return new TableFunctionProcessorProvider()
+        @Override
+        public TableFunctionProcessorProvider getTableFunctionProcessorProvider(ConnectorTableFunctionHandle functionHandle)
         {
-            @Override
-            public TableFunctionSplitProcessor getSplitProcessor(ConnectorSession session, ConnectorTableFunctionHandle handle, ConnectorSplit split)
-            {
-                return new SequenceFunctionProcessor(((SequenceFunctionHandle) handle).step(), (SequenceFunctionSplit) split);
+            if (functionHandle instanceof SequenceFunctionHandle) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/table/Sequence.java:260]: invert `if` to reduce overall code indentation

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java (hunk 3 above)
Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/metadata/TableFunctionProcessorProviderFactory.java]: moved to single `TableFunctionFeaturesFactory` interface,
but I am not sure about the name

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionFeaturesFactory.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package io.trino.metadata;
+
+import io.trino.spi.connector.ConnectorSplitSource;
+import io.trino.spi.function.table.ConnectorTableFunctionHandle;
+import io.trino.spi.function.table.TableFunctionProcessorProvider;
+
+import java.util.Optional;
+
+public interface TableFunctionFeaturesFactory
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionFeaturesFactory.java:22]: How about 

```
interface TableFunctionProvider {
   Optional<TableFunction> get(ConnectorSession session, ConnectorTableFunctionHandle functionHandle);
}

interface  TableFunction {
  TableFunctionProcessorProvider getProcessorProvider(ConnectorSession session);
  ConnectorSplitSource getSplitSource(ConnectorSession session);
}
```

However I am not sure about `TableFunction` name, it is too generic.

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 7]
```diff
@@ -15,7 +15,7 @@
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:15]: Improve the commit message. In such case I would write `Introduce X` and then explain why X is beneficial

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 8]
```diff
@@ -93,11 +95,12 @@ public ConnectorSplitManager getSplitManager()
...
-                if (functionHandle instanceof SequenceFunctionHandle sequenceFunctionHandle) {
-                    return getSequenceFunctionSplitSource(sequenceFunctionHandle);
-                }
-
-                throw new UnsupportedOperationException();
+                return tableFunctionFeaturesFactories.stream()
+                        .map(factory -> factory.getSplitSource(functionHandle))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .findFirst()
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:102]: What if the are many matches? I think you shoulld use `toOptional` collector.

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 9]
```diff
@@ -93,11 +95,12 @@ public ConnectorSplitManager getSplitManager()
...
+                        .orElseThrow(UnsupportedOperationException::new);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:103]: I think we need a better exception than that.

**Review by kokosing — COMMENTED** [review]
I like this refactor.

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionFeaturesFactory.java (hunk 6 above)
Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/metadata/TableFunctionFeaturesFactory.java]: Refactored, but can not find proper naming

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java (hunk 8 above)
Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:111]: check size added, we should have only zero or one

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java (hunk 9 above)
Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:112]: took it from previous code

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 10]
```diff
@@ -15,7 +15,7 @@
 
 import com.google.common.collect.ImmutableSet;
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:16]: update commit message, there is no TableFunctionAdditionalFeaturesProvider any more

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionProvider.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+     * with TableFunctionProcessorProvider and ConnectorSplitSource implementation
+     * for concrete function handle
+     */
+    Optional<TableFunction> get(ConnectorTableFunctionHandle functionHandle);
+
+    interface TableFunction
+    {
+        default TableFunctionProcessorProvider getTableFunctionProcessorProvider()
+        {
+            return null;
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionProvider.java:35]: I don't think `return null` is a good default implementation. I would make this method abstract (without implementation). I think it is is fine to `throw UnsupportedOperationException` by default if it is ok to not implement such function. Then we could learn that method was called even if it was expected. However, returning `null` is bad as a failure may happen later when the result would be actually consumed.

**Comment by kokosing** [comment]
Is this still a draft?

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionProvider.java (hunk 11 above)
Comment (vlad-lyutenko) [reply core/trino-main/src/main/java/io/trino/metadata/TableFunctionProvider.java]: > I think it is is fine to `throw UnsupportedOperationException`

made default to throw `UnsupportedOperationException`

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 12]
```diff
@@ -46,13 +49,15 @@ public class GlobalSystemConnector
     private final Set<SystemTable> systemTables;
     private final Set<Procedure> procedures;
     private final Set<ConnectorTableFunction> tableFunctions;
+    private final Set<TableFunctionProvider> tableFunctionProviders;
 
     @Inject
-    public GlobalSystemConnector(Set<SystemTable> systemTables, Set<Procedure> procedures, Set<ConnectorTableFunction> tableFunctions)
+    public GlobalSystemConnector(Set<SystemTable> systemTables, Set<Procedure> procedures, Set<ConnectorTableFunction> tableFunctions, Set<TableFunctionProvider> tableFunctionProviders)
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:55]: nit: Can we swap the arguments ?

Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:55]: Are we going to replace the `tableFunctions` to `tableFunctionProvider`

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 13]
```diff
@@ -93,11 +98,18 @@ public ConnectorSplitManager getSplitManager()
...
             public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, ConnectorSession session, ConnectorTableFunctionHandle functionHandle)
             {
-                if (functionHandle instanceof SequenceFunctionHandle sequenceFunctionHandle) {
-                    return getSequenceFunctionSplitSource(sequenceFunctionHandle);
-                }
+                Set<TableFunction> tableFunctionAdditionalFeatures = tableFunctionProviders.stream()
+                        .map(factory -> factory.get(functionHandle))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .collect(Collectors.toSet());
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:105]: nit: ImmutablSet ?

File: core/trino-main/src/main/java/io/trino/connector/system/SystemConnectorModule.java [hunk 14]
```diff
@@ -77,6 +80,8 @@ public void configure(Binder binder)
         binder.bind(KillQueryProcedure.class).in(Scopes.SINGLETON);
 
         binder.bind(GlobalSystemConnector.class).in(Scopes.SINGLETON);
+        Multibinder<TableFunctionProvider> tableFunctionAdditionalFeaturesProviders = Multibinder.newSetBinder(binder, TableFunctionProvider.class);
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/connector/system/SystemConnectorModule.java:83]: We can rename the variable as `tableFunctionProviders`

File: core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java [hunk 15]
```diff
@@ -42,27 +42,42 @@
...
         implements FunctionProvider
 {
     public static final String BUILTIN_SCHEMA = "builtin";
     private volatile FunctionMap functions = new FunctionMap();
+    private final Set<TableFunctionProvider> tableFunctionProviders;
+
+    public GlobalFunctionCatalog()
+    {
+        this(ImmutableSet.of());
+    }
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java:74]: Can we modify the users of `GlobalFunctionalCatalog` to use constructor one with functionProviders ?

File: core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java [hunk 16]
```diff
@@ -93,11 +98,18 @@ public ConnectorSplitManager getSplitManager()
...
-                    return getSequenceFunctionSplitSource(sequenceFunctionHandle);
-                }
+                Set<TableFunction> tableFunctionAdditionalFeatures = tableFunctionProviders.stream()
+                        .map(factory -> factory.get(functionHandle))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .collect(toImmutableSet());
 
-                throw new UnsupportedOperationException();
+                checkArgument(tableFunctionAdditionalFeatures.size() <= 1, "we should have only one implementation for concrete function handle or nothing");
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/connector/system/GlobalSystemConnector.java:107]: This should be a TrinoException with `FUNCTION_IMPLEMENTATION_ERROR` error code. Also, rephrase as "multiple implementations found for a function handle".

**Comment by Praveen2112** [comment]
> What's the motivation for this change?

The goal is to simplify the registration of a function which is closer to the engine like if we are having a different table function like `SequenceFunctionHandle` we could need to introduce another `if` statement validating for the `functionHandle` implementation. 

Is there is any other approaches we could take to simplify the table function registration.

**Comment by martint** [comment]
As I described above, we should avoid adding functions in the global namespace and, instead, implement them via connectors that get mounted in a catalog/schema. It's very likely that in the future we'll move those existing functions to a connector.

**Comment by mosabua** [comment]
@vlad-lyutenko I assume you are still continuing on this work. If not, please close the PR.

**Pull request closed without merging**